## 1.1 The Elements of Programming

Programming languages have three mechanisms for combining simple ideas to form more complex ideas:
1. *primitive expressions*, the simplest entities the language is concerned with
2. *means of combination* by which compound elements are built from simpler ones, and
3. *means of abstraction* by which compound elements can be named and manipulated as units

Programming deals with two kinds of elements: *functions* and *data* (but later we will discover they are really not so distinct).
Informally,  
1. data: stuff we want to manipulate
2. functions: descriptions of rules for manipulating data

### 1.1.1 Expressions

We start by typing *statements* into the *interpreter* that *evaluates* those statements.  

We start with an expression statement: in javascript, an *expression* followed by a semicolon. One kind of primitive expression is a number.

In [1]:
439;

439

Expressions representing numbers may be combined with *operators* such as `+` or `*` to form a *compound expression* that represents the application of a corresponding *primitive function* to those numbers:

In [2]:
137 + 439;

576

Expressions such as these, which contain other expressions as components, are called *combinations*.

When a combination is formed by an *operator* symbol in the middle, and *operand* expressions to the left and right, we call it an *operator combination*. The value of an operator combination is obtained by applying the function specified by the operator to the arguments that are the values of the operands.

We place the operator between the operands by convention, a convention known as *infix notation*. It follows mathematical notation used in school and everday life.

As in mathematics, operator combiniations can be *nested*, that is, they can have operands that themselves are operator combinations:

In [3]:
(3 * 5) + (10 - 6);

19

Parentheses are used to group operator combinations, in order to avoid ambiguities.

Javavscript also follows the conventions that multiplication and division bind more strongly than addition and subtraction, i.e, `3 * 5 + 10 / 2;` is equivalent to `(3 * 5) + (10 / 2);`. We say that `*` and `/` have *higher precedence* than `+` and `-`.

Sequences are read from left to right. Thus

```javascript
1 - 5 / 2 * 4 + 3;
```

stands for

```javascript
(1 - ((5 / 2) * 4)) + 3;
```

We say that these operators are *left-associative*.

Lastly, note that, regardless of the complexity of the expression, the interpreter always operates in the same simple cycle:
1. read a statement typed by the user
2. evauate the statement
3. print the result

We refer to this mode of operation as a *read-evaluate-print loop*, or the initialism *REPL*.

### 1.1.2 Naming and the Environment

A critical aspect of a programming language is the means it provides for using names to refer to computational objects.
We say the name identifies a constant whose *value* is the object.

The first such means we will learn in JavaScript is a *constant*. In JavaScript, we name constants with *constant declarations*.

In [4]:
const size = 2;

Once the name `size` has been associated with thbe number 2, we can refer to the value 2 by name:

In [5]:
size;

2

In [6]:
5 * size;

10

Constant declaration is our language's simplest means of abstraction, for it allows us to use simple names to refer to the results of compound operations, like so:

In [7]:
const pi = 3.14159;

const radius = 10;

const circumference = 2 * pi * radius;

circumference;

62.8318

Means of abstraction like this allow us to build up computational objects with very complex structures. We write programs to avoid having to remember and repeat the details of building up these objects every time we want to use them. Instead, we write programs that construct computational objects of increasing complexity, by building up from simple expressions, step by step.

The fact that we can associate names with values implies that the interpreter must maintain some sort of memory that keeps track of name-object pairs. This memory is called the *environment* -- more precisely, the *program environment*, as a computation may involve a number of different environments.

### 1.1.3 Evaluating operator combinations

One goal of this chapter is to isolate issues about thinking procedurally.

As a case in point, let us consider the procedure the interpreter follows when evaluating operator combinations.

* To evaluate an operator combination, do the following:
  1. Evaluate the operand expressions of the combination.
  2. Apply the function that is denoted by the operator to the arguments that are the values of the operands.

Even this simple rule illustrates some important points about procedures.  
First, observe that the first step dicates we must first perform the evaluation process on each operand. Thus, the evaluation rule is *recursive*: it includes, as one of its steps, the need to invoke the rule itself.

Recursion allows us to express succinctly what might otherwise appear to be a rather complicated process.  
For example, the following expression requires that the evaluation rule be applied to four different combinations:

In [8]:
(2 + 4 * 6) * (3 + 12);

390

We can picture this process by representing the combination in the form of a tree, as shown in figure 1.1. Each combination is represented by a *node* with *branches* corresponding to the operator and the operands of the combination stemming from it. The *terminal* nodes, with no branches stemming from them, represent either operators or numbers.

![](./figs/fig1.1.png)

If we view evaluation in terms of the tree, we can imagine that the values of the operands percolate upward, starting from the terminal nodes and then combining at higher and higher levels. **In general, we will see that recursion is a very powerful technique for dealing with hierarchical, treelike objects.** In fact, the "percolate values upward" form of the evaluation rule is an example of a general kind of process known as *tree accumulation*.

The second point we can observe about this procedure is that our rules don't tell us how to evaluate primitve expressions -- the *base case* of our recursion. We take care of the primitive cases by stipulating that. 
* the values of numerals are the numbers that they name, and
* the values of names are the objects associated with those names in the environment.

The key point here is that **the enviornment plays a role in determining the meaning of names in expressions**. In an interpreted language such as JavaScript, especially when run in an interactive REPL, it is meaningless to talk about the value of an expression such as `x + 1;` without specifying any information about what value the environment would provide for `x`.

#### Declarations are not combinations
Notice the evaluation rule above does not handle declarations, like the constant declarations we saw above.  
We should not think of evaluating `const x = 3;` as applying an equality operator to two arguments `x` and `3`. Therefore `const x = 3;` is not a combination.
#### Keywords and rules

The word `const` is a *keyword* in JavaScript, a word with a particular meaning. Because they carry these meanings, keywords in languages are always reserved and cannot be used as names (at least, not without breaking programs).

A keyword or combination of keywords in a statement instructs the interpreter to treat the statement in a special way. **Each such *syntactic form* has its own evaluation rule.**

### 1.1.4 Compound Functions

So far we have demonstrated with JavaScript some of the elements common to many programming languages:
* primitive data and functions: numbers and arithmetic operations
* means of combining operations, e.g., nesting with parentheses
* a limited means of abstraction: constant declarations that associate names with values

Next we learn about a much more powerful abstraction technique: *function declarations* that associate a name with a compound operation, so that the operation can be referred to as a unit.

We being with the idea of squaring a number. To express this idea in plain language, we might say, "To square something, take it times itself." We express this in JavaScript as:

In [9]:
function square(x) {
    return x * x;
}

The name `square` identifies our *compound function*. The function represents the operation of multiplying something by itself. Notice that the thing to be multiplied is given a local name, `x`. This is similar to the role pronouns play in natural language.

The simplest form of a function declaration is:

`function` *name*`{`*parameters*`) { return` *expression*`; }`

where *name* is a symbol to be associated with the function in the environment, and the *parameters* are the names used within the body of the function to refer to the corresponding arguments to the function.

Here we only have one parameter, but more generall we have a comma-separated list of a parameters, and in an application of the function we will likewise have a comma-separated list of arguments, the values that the parameters take on when the function is applied.

In its simplest form, the *body* of a function declaration is a single *return statement*, the keyword `return` followed by the *return expression* that will yield the value of the function application.

Having declared `square`, we can now use it in a *function application* expression.

In [10]:
square(21);

441

Function applications are--after operator combinations--the second kind of combination of expressions into larger expressions that we encounter. The general form is:

*function-expression(argument-expressions)*

where the *function-expression* of the application specifies the function to be applied to the comma-separated *argument-expressions*.

To evaluate a function application, the interpreter follows a procedure quite similar to the procedure for operator combinations described in section 1.1.3.

* To evaluate a function application, do the following:
  1. Evaluate the subexpressions of the application, namely the function expression and the argument expressions.
  2. Apply the function that is the value of the function expression to the values of the argument expressions.

To see the generality of this procedure, take for example a case where the argument expression is itself a compound expression:

In [11]:
square(2 + 5);

49

We can see that the interpreter indeed follows our procedure, evaluating the argument expression first before applying the value of the function expression to the value of the argument expression.

Observe that function applicaton expressions can also serve as argument expressions:

In [12]:
square(square(3));

81

We can also use `square` as a building block in defining other functions. For example, $x^2 + y^2$ can be expressed as:

`square(x) + square(y)`

Using this expression, we can declare a functio `sum_of_squares` that, given any two numbers as arguments, produces the sum of their squares:

In [13]:
function sum_of_squares(x, y) {
    return square(x) + square(y);
}

In [14]:
sum_of_squares(3, 4);

25

And by extension we can use `sum_of_squares` as a building block in constructing yet more functions:

In [15]:
function f(a) {
    return sum_of_squares(a + 1, a * 2);
}

In [16]:
f(5);

136

In addition to compound functions, any JavaScript environment provides primitive functions that are built into the interpreter or loaded from libraries. These additional primitive functions are used in the exact same way as compuond functions. This means that it doesn't matter when we write a compound function like `sum_of_squares` whether `square` was built into the interpreter, loaded from a library, or itself defined as a compound function.

### 1.1.5 The Substitution Model for Function Application

To evaluate a function application, the interpreter follows the procedure described in section 1.1.4. For primitive functions, this is handled by the interpreter or libraries. For compound functions, the application process is as follows:
* To apply a compound function to arguments, evaluate the return expression of the function with each parameter replaced by the corresponding argument.

We can apply this process to the functions we declared above.  
Let's evaluate the application

```javascript
f(5);
```

where `f` is the function we just declared.  
We begin by retrieving the return expression of `f`:
```javascript
sum_of_squares(a + 1, a * 2);
```
Then we replace the parameter `a` with the argument 5:
```javascript
sum_of_squares(5 + 1, 5 * 2)
```
Thus the problem reduces to the evaluation of an application with two arguments and a function expression `sum_of_squares`. Evaluating this applications involves three subproblems. We continue substituting to reduce these subproblems until we end up with two primitive expressions and a primitive function, an operator applied to them: `36 + 100`. This gives us, finally, `136`.

The process we have just described is called the *substitution model* for function application.

Two points that should be stressed:
1. The purpose of the substitution is to help us to think about function application. It is *not* a description of how the interpreter really works: typically interpreters do not evaluate function applications by manipulating the text of the function to substitute values for parameters. In practice, the "substitution" is accomplished by using a local environment for the parameters.
2. Over the course of this book, we present a sequence of increasingly elaborate models of how interpreters work, culminating with a complete implementation of an interpreter and compiler. The substitution model is only the first of these. When we address the use of functions with mutable data, we will see that the substitution model breaks down and must be replaced by a more complicated model of function evaluation.

#### Applicative order versus normal order

The procedure given in 1.1.4 is not the only way to perform evaluation. An alternative model would not evaluate arguments until their values were needed. Instaed, it would first substitute argument epxressions for parameters until it obtained an expression involving only operators and primitive functions, and would then perform the evaluation.

If we used this method, the evaluation of `f(5)` would proceed according to the sequence of expansions
```javascript
sum_of_squares(5 + 1, 5 * 2)

square(5 + 1) + square(5 * 2)

(5 + 1) * (5 + 1) + (5 * 2) * (5 * 2)
```
followed by the reductions
```javascript
6 * 6 + 10 * 10
36 + 100
136
```
This gives the same answer as our previous evaluation model, but the process is different. In particular, the evaluations of `5 + 1` and `5 * 2` are each performed twice here, corresponding to the reduction of the expression
```javascript
x * x
```
with `x` replaced, respectively, by `5 + 1` and `5 * 2`.

With the order we followed above, we would isntead have done
```javascript
sum_of_squares(5 + 1, 5 * 2)

sum_of_squares(6, 10)

square(6) + square(10)
```
and then the reduction would have proceeded as just shown.

This alternative method of fully expanding and then reducing is known as *normal-order evaluation*. The interpreter actually evaluates the arguments and then applies functions, and this is called *applicative-order evaluation*. It can be shown that, for function applications that can be modeled using substitution and that yield legitimate values, normal-order and applicative-order evaluation produce the same value. JavaScript uses applicative-order evaluation. This is partly because of the additional efficiency obtained from avoiding multiple evauations of expressions, as just demonstrated. More significantly, normal-order evaluation becomes much more complicated to deal with when we leave the realm of functions that can be modeled by substitution. On the other hand, normal-order evaluation can be an extremely valuable tool.

### 1.1.6 Conditional Expressions and Predicates

The expressive power of the class of functions that we can define at this point is very limited, because we have no way to make tests and to perform different operations depending on the result of a test.

$$
|x| = 
\begin{cases}
x \text{ if } x \ge 0 \\
-x \text{ otherwise }
\end{cases}
$$

This construct is a *case analysis* and can be written in JavaScript using a *conditional expression* as

In [17]:
function abs(x) {
    return x >= 0 ? x : -x;
}

This can be expressed in English as "If $x$ is greater than or equal to zero, return $x$; otherwise return $-x$".

The general form of a conditional expression is 

*predicate* `?` *consequent-expression* `:` *alternative-expression*

Conditional expressions begin with a *predicate*, an expression whose value evaluates to either *true* or *false*, the two distinguished *boolean* values in JavaaScript. The primitive boolean expressions are `true` and `false`. The predicate is followed by a question mark, then the *consequent-epxression*, a colon, and finally the *alternative-expression*.

To evaluate a conditional expression, the interpreter starts by evaluating the *predicate*. If the *predicate* evaluates to true, the interpreter evaluates the *consequent-expression* and returns its value as the value of the conditional. If the predicate evaluates as false, the interpreter evaluates the *alternative-expression* and returns that instead.

The word *predicate* is used for operators and functions that return true or false, as well as for expressions that evaluate to true or false. Our absolute-value function `abs` makes use of the primitive predicate `>=`.

We could specify our absolute value function so that it has more than two cases:
$$
|x| = 
\begin{cases}
&x &\text{if } x \gt 0 \\
&0 &\text{if } x = 0 \\
&-x &\text{otherwise}
\end{cases}
$$

In JavaScript, we express a case analysis with multiple cases by nesting conditional expressions as alternative experssions inside other conditional expressions:

In [18]:
function abs(x) {
    return x > 0
           ? x
           : x === 0
           ? 0
           : -x;
}

Parentheses are not needed around the alternatve expression `x === 0 ? 0 : -x` because the conditional-expression syntactic form is right-associative.

The general form of a case analysis is  
$p_1$
`?` $e_1$  
`:` $p_2$  
`?` $e_2$  
$\vdots$  
`:` $p_n$  
`?` $e_n$  
`:` *final-alternative-expression*

We call a predicate $p_i$ together with its consequent expression $e_i$ a *clause*. A case analysis can be seen as a sequence of clauses, followed by a final alternative expression. We can evaluate a case analysis using what we know about how the interpreter evaluates conditonal expressions: first the interpreter evaluates the predicate $p_1$. If its value is false, then $p_2$ is evaluated. If $p_2$'s value is also found to be false, then $p_3$ is evaluated. this process continues until a predicate $p_i$ is found whose value is true, in which case the interpreter returns the value of the corresponding consequent expression $e_i$ as the value of the case analysis. If none of the $p$s is found to be true, the value of the case analysis is the value of the final alternative expression.

In addition to primitve predicates (`>=`, `>`, `<`, `<=`, `===`, and `!==`) that are applied to numbers, there are logical composition operations that enable us to construct *compound predicates*. The three most frequently used are these:

1. $expression_1$ `&&` $expression_2$  
   This operation expresses *logical conjunction*, roughly the same as the English word "and". We assume this form to be syntactic sugar for $expression_1$ `?` $expression_2$ `: false`

2. $expression_1$ `||` $expression_2$  
   This operation expresses *logical disjunction*, roughly the same as the English word "or". We assume this form to be syntactic sugar for $expression_1$ `?` `true :` $expression_2$

3. `!` $expression$  
   This operation expresses *logical negation*, roughly the same as the English word "not". The value of the expression is true when $expression$ evaluates to false, and false when $expresssion$ evaluates to true.

In [ ]:

s